File: pointpub-sample-objc/pointpub-sample-objc/ViewController.m
//
//  ViewController.m
//  pointpub-sample-objc
//
//  Created by neptune on 9/8/25.
//

#import "ViewController.h"
#import "PointPubSDK/PointPubSDK-Swift.h"

@interface ViewController () <PointPubDelegate>
@property (weak, nonatomic) IBOutlet UIButton *enterButton;
@property (weak, nonatomic) IBOutlet UIButton *getVirtualPointButton;
@property (weak, nonatomic) IBOutlet UIButton *spendVirtualPointButton;
@property (weak, nonatomic) IBOutlet UIButton *getCampaignButton;

@end

@implementation ViewController

// NSString *appId = "<ENTER YOUR APP ID>"
NSString *appId = @"APP_17569663893761798";
// NSString *userId = "<ENTER YOUR USER ID>"
NSString *userId = @"123456789";

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UITapGestureRecognizer *tapGesture = [
        [UITapGestureRecognizer alloc] initWithTarget: self action: @selector(didTapView)
    ];
    [self.view addGestureRecognizer:tapGesture];
    
    _enterButton.layer.cornerRadius = 12;
    _getVirtualPointButton.layer.cornerRadius = 12;
    _spendVirtualPointButton.layer.cornerRadius = 12;
    _getCampaignButton.layer.cornerRadius = 12;
    
    PointPub.delegate = self;
    [PointPub setAppIdWith: appId];
    [PointPub setUserIdWith: userId];
    if (!PointPub.isTrackingEnabled) {
        [PointPub requestTrackingPermission:^(BOOL isTrackingPermission) {
            
        }];
    }
}

- (void)didTapView {
    [self.view endEditing:YES];
}

- (IBAction)didTapEnterButton:(id)sender {
    [PointPub startOfferwallFrom: self];
}

- (IBAction)didTapGetVirtualPoint:(id)sender {
    [PointPub getVirtualPoint:^(NSString *name, NSNumber *point, NSError *error) {
        if (error) {
            NSLog(@"❌ Error: %@", error.localizedDescription);
        } else {
            NSLog(@"✅ 포인트명: %@, 포인트 잔액: %ld", name, [point longValue]);
        }
    }];
}

- (IBAction)didTapSpendVirtualPoint:(id)sender {
    [self spendAllVirtualPointsIfAvailable];
}

- (void)spendAllVirtualPointsIfAvailable {
    [PointPub getVirtualPoint:^(NSString *name, NSNumber *remainingPoint, NSError *error) {
        if (error) {
            NSLog(@"❌ Error: %@", error.localizedDescription);
        } else {
            NSLog(@"✅ 포인트명: %@, 포인트 잔액: %ld", name, [remainingPoint longValue]);
            
            if ([remainingPoint intValue] > 0) {
                [self spendVirtualPoint:remainingPoint];
            }
        }
    }];
}

- (void)spendVirtualPoint:(NSNumber*)amount {
    [PointPub spendVirtualPointWithPoint:amount:^(NSString *name, NSNumber *point, NSError *error) {
        if (error) {
            NSLog(@"❌ Error: %@", error.localizedDescription);
        } else {
            NSLog(@"✅ 포인트명: %@, 포인트 잔액: %ld", name, [point longValue]);
        }
    }];
}

- (IBAction)didTapGetCompletedCampaign:(id)sender {
    [PointPub getCompletedCampaign:^(NSString *response, NSError *error) {
        if (error) {
            NSLog(@"❌ Error: %@", error.localizedDescription);
        } else {
            NSLog(@"✅ response: %@", response);
        }
    }];
}

- (void)onOpenOfferwall {
    NSLog(@"onOpenOfferwall");
}

- (void)onCloseOfferwall {
    NSLog(@"onCloseOfferwall");
}

@end

File: pointpub-sample-swift/pointpub-sample-swift/ViewController.swift
//
//  ViewController.swift
//  pointpub-sample-swift
//
//  Created by neptune on 9/8/25.
//

import UIKit
import PointPubSDK

class ViewController: UIViewController {
    
    @IBOutlet weak var enterButton: UIButton!
    @IBOutlet weak var getVirtualPointButton: UIButton!
    @IBOutlet weak var spendVirtualPointButton: UIButton!
    @IBOutlet weak var getCampaignButton: UIButton!
    
    // let appId = "<ENTER YOUR APP ID>"
    private final let appId = "APP_17569663893761798"
    // let appId = "<ENTER YOUR USER ID>"
    private final let userId = "123456789"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(didTapView))
        view.addGestureRecognizer(tapGesture)
        
        enterButton.layer.cornerRadius = 12
        getVirtualPointButton.layer.cornerRadius = 12
        spendVirtualPointButton.layer.cornerRadius = 12
        getCampaignButton.layer.cornerRadius = 12
        
        PointPub.delegate = self
        PointPub.setAppId(with: appId)
        PointPub.setUserId(with: userId)
        
        if !PointPub.isTrackingEnabled() {
            PointPub.requestTrackingPermission { isTrackingPermission in
                
            }
        }
    }
    
    @objc
    private func didTapView(_ sender: Any) {
        view.endEditing(true)
    }

    @IBAction func didTapEnterButton(_ sender: Any) {
        PointPub.startOfferwall(from: self)
    }
    
    @IBAction func didTapGetVirtualPoint(_ sender: Any) {
        Task {
            do {
                let (pointName, point) = try await PointPub.getVirtualPoint()
                print("포인트명: \(pointName), 포인트 잔액: \(point)")
            } catch {
                print(error.localizedDescription)
            }
        }
    }
    
    @IBAction func didTapSpendVirtualPoint(_ sender: Any) {
        spendAllVirtualPointsIfAvailable()
    }
    
    private func spendAllVirtualPointsIfAvailable() {
        Task {
            do {
                let (pointName, remainingPoint) = try await PointPub.getVirtualPoint()
                print("포인트명: \(pointName), 포인트 잔액: \(remainingPoint)")
                    
                if remainingPoint > 0 {
                    spendVirtualPoint(amount: remainingPoint)
                }
            } catch {
                print(error.localizedDescription)
            }
        }
    }

    private func spendVirtualPoint(amount: Int) {
        Task {
            do {
                let (pointName, remainingPoint) = try await PointPub.spendVirtualPoint(point: amount)
                print("포인트명: \(pointName), 사용 후 포인트 잔액: \(remainingPoint)")
            } catch {
                print(error.localizedDescription)
            }
        }
    }
    
    @IBAction func didTapGetCompletedCampaign(_ sender: Any) {
        Task {
            do {
                let response = try await PointPub.getCompletedCampaign()
                print("\(response)")
            } catch {
                print(error.localizedDescription)
            }
        }
    }
}

extension ViewController: PointPubDelegate {
    func onOpenOfferwall() {
        print(#function)
    }
    
    func onCloseOfferwall() {
        print(#function)
    }
}
